0006-ffmul-revised.md

```diff
@@ -0,0 +1,1932 @@
+
+#### High limb checks
+
+Given that all the limbs of $a, b, q, r$ are checked to be correct (at most $\ell$ bits each), the CRT bound can be satisfied if we just check further that their high limbs are correctly upper bounded like:
+
+$$x'_2 = x_2 + 2^\ell - f_2 - 1 < 2^\ell$$
+
+which is equivalent to saying that 
+
+$$x_2 \leq f_2$$
+
+This saves us $2.66$ range-check rows per term above. Moreover, it prevents us from computing the full bounds, saving us a considerable number of rows in terms of `ForeignFieldAdd` gadgets.
+
+Either way, the high limbs of the input terms can be checked somewhere else, and they only contribute to at most $1.33\cdot 2$ rows per `ForeignFieldMul` (as opposed to 8). Plus, the high bounds computation itself can be carried out in one single `Generic` gate, taking $1$ row as opposed to $2\cdot 2$ rows before (using foreign field addition).
+
+Similarly, the approach benefits the elements which are "newly" created within the `ForeignFieldMul` gate. That is, the quotient and remainder. 
+
+#### Bounding the quotient
+
+Given that we have sufficient space left in the gate, we can compute the actual high limb bound of one extra term within the `ForeignFieldMul` (instead of using a `Generic` for it). Since the quotient is an internal witness and is not meant to be used externally, we will include the high bound of $q$ inside (saving us $0.5$ rows). 
+
+$$
+q'_{2} = q_2 + 2^{\ell} - f_2 - 1
+$$
+
+Then, this value can be wired to a single-limb range check, using only $1.33$ rows. Because the constraints of the gate were using one multi range check gadget to constrain $p_{10}$ and $p_{110}$, it makes sense to use that remaining space to constrain $q'_2$ as well. 
+
+#### Remainder compact limbs
+
+Due to the limited number of permutable cells per gate, we do not have enough cells for copy constraining $q'$ and $r'$ (or $q$ and $r$) to their respective range check gadgets. To address this issue, we must decompose $r$ into 2 limbs instead of 3. This is doable because the bottom part of $r$ is always used together in the constraints (this is not doable in $q$ however):
+
+$$
+r = r_{01} + 2^{2\ell} \cdot r_2
+$$
+
+Then, $r_2$ is stored in a permutable cell and can be used by a _half_ `Generic` gate to compute 
+
+$$
+r'_2 = r_2 + 2^{2\ell} - f_2 - 1
+$$
+
+Note that $r'_2$ must be range checked by a `multi-range-check` gadget. Then, the exteral checks structure will keep track of it and make "dense" multi range check with triples of remainders coming from a number of `ForeignFieldMul` from a larger circuit. 
+
+- Store a copy of the limbs $r_{01}$ and $r_2$ in its witness
+- Compact-range check that each limb is $\ell$ bits each
+- Decompose it into $r_0, r_1, r_2$ in the compact range check gadget
+- Compute $r'_2$ using a `Generic` gate
+- Push the bound to a list of external checks to be multi range checked together.
+
+
+### Inlining compact full bound check inside a gate
```

Comment:
This feels like it would be better suited as a separate, dedicated document

Reply:
Deleting for now